### PR TITLE
testing Array.prototype

### PR DESCRIPTION
**Pull request by anvidmen**

### CODE REVIEW

File: staff/juandavid-omen/arrays/array-functions/every.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    if (array.length ==0){
+        return false;
+    }
+    var counter = 0;
+    for (var i = 0; i < array.length; i++) {
+        if(expression(array[i], i, array)){
+            counter +=1;
+        };
+    }
+    if(counter === array.length){
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/every.js:11]: do you really need to iterate the full array to detect if all elements match? think about it

File: staff/juandavid-omen/arrays/array-functions/filter.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+
+function filter(array, expression) {
+    var result = [];
+
+    for (let i = 0; i < array.length; i++) {
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/filter.js:5]: use var, not let. we are on es5 mode.

File: staff/juandavid-omen/arrays/array-functions/find.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+
+function find(array, expression) {
+    for (let i = 0; i < array.length; i++) {
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/find.js:3]: use var, not let.

File: staff/juandavid-omen/arrays/array-functions/find.spec.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+describe('The find method', function () {
+    it('returns the value of the first element in the provided array that satisfies the provided testing function' , function () {
+        var array = [5, 12, 8, 130, 44];
+        var result = find(array, element => element > 10);
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/find.spec.js:4]: arrow functions are not allowed. we are on es5 mode. move it to a normal function.

File: staff/juandavid-omen/arrays/array-functions/find.spec.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    });
+    it('returns undefined if no elements in the provided array satisfy the provided testing function' , function () {
+        var array = [5, 12, 8, 130, 44];
+        var result = find(array, element => element > 200);
+
+        expect(result).toBe(undefined);
+    });
+    it('returns undefined if the length of the array is 0' , function () {
+        var array = [];
+        var result = find(array, element => element > 0);
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/find.spec.js:16]: move arrow functions to normal functions.

File: staff/juandavid-omen/arrays/array-functions/findIndex.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+function findIndex(array, expression) {
+    for (let i = 0; i < array.length; i++) {
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/findIndex.js:2]: use var, not let.

File: staff/juandavid-omen/arrays/array-functions/indexOf.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+function indexOf(array, element, index){
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/indexOf.js:1]: rename the file from indexOf.js to index-of.js. remember: file names always in lower-case (with dash between words).

File: staff/juandavid-omen/arrays/array-functions/indexOf.spec.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+describe('indexOf', function () {
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/indexOf.spec.js:1]: rename this file to index-of.spec.js

File: staff/juandavid-omen/arrays/array-functions/pop.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+function pop(array) {
+    
+    if (array.length-1 < 0) {
+        return undefined;
+    }
+    var lastElement = array[array.length-1]
+    array.length -= 1;
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/pop.js:7]: may you use de decrement operator instead (--)

File: staff/juandavid-omen/arrays/array-functions/push.js [hunk 10]
```diff
@@ -0,0 +1,6 @@
+function push(array, element){
+    for (var i = 1; i<arguments.length;i++){
```
Comment (manuelbarzi) [inline staff/juandavid-omen/arrays/array-functions/push.js:2]: you using arguments, ok, but then why you declare de parameter element? your are no using it, then remove it.

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.join.spec.js [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    expect(result).toBe("5,true,function hello(){}");
+  });
+
+  it("throw an error when the separator is not a string", function () {
+    var array = new Arroz(5, 7, 8);
+
+    expect(function () { 
+      array.join(1);
+     }).toThrowError(TypeError, '1 is not a string');
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.join.spec.js:66]: same

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.push.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+"use strict";
+
+Arroz.prototype.push = function() {
+  for (var i in arguments) {
+    if (typeof arguments[i] !== 'number') {
+      throw new TypeError(arguments[i] + ' must be numeric');
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.push.js:6]: Same here, and this is a bit more serious, you should know that you can push anything into an array. Here you are only letting the user push numbers

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.push.spec.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+      expect(array[1]).toBe(7);
+      expect(array[2]).toBe(8);
+    });
+
+  it("should throw an error when not numeric arguments", function () {
+    var array = new Arroz();
+
+    expect(function () {  
+      array.push(true, 'e', '8');
+    }).toThrowError(TypeError, 'true must be numeric');
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.push.spec.js:41]: same

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.reduce.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
+"use strict";
+
+Arroz.prototype.reduce = function(expression, initialValue) {
+    if (this.length === 0) {
+        throw new TypeError('Arroz has not values to do reduce');
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.reduce.js:5]: this error only happens if you don't provide initial value, if you do it return 0 so no error should be thrown

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.reduce.spec.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  });
+
+  it('Arroz.prototype.reduce() Errors', function () {
+    var array = new Arroz();
+    
+    expect(function () {
+      array.reduce(function (accumulator, element) {
+        return accumulator + element;
+      });
+    }).toThrowError(TypeError, 'Arroz has not values to do reduce');
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.reduce.spec.js:53]: This error as I said before does not happen in the real Array

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.slice.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+    if(arguments.length == 2) {
+        begin = arguments[0];
+        end = arguments[1];
+    } else if(arguments.length == 1) {
+        end = arguments[0];
+    }
+    
+    if (begin > this.length) {
+        throw new RangeError(begin + " is higher than the index range");
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.slice.js:16]: This is not how slice method works

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.slice.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    if (end > this.length) {
+        throw new RangeError(end + " is higher than the index range");
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.slice.js:20]: same here

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.slice.spec.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
+"use strict";
+
+describe("Arroz.prototype.slice", function() {
+  it("should throw an error when the begin param is higher than the index range", function() {
+    var array = new Arroz(5, 12, 8, 130, 44);
+    
+    expect(function() {
+      array.slice(6, 10);
+    }).toThrowError(RangeError, "6 is higher than the index range")
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.slice.spec.js:9]: this should be removed as per the comments I made above

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.slice.spec.js [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    
+  });
+
+  it("should throw an error when the end param is higher than the index range", function () {
+    var array = new Arroz(5, 12, 8, 130, 44);
+
+    expect(function () {
+      array.slice(1, 10);
+    }).toThrowError(RangeError, "10 is higher than the index range")
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.slice.spec.js:18]: same here

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.splice.js [hunk 20]
```diff
@@ -0,0 +1,50 @@
+"use strict";
+
+Arroz.prototype.splice = function(start, deleteCount) { 
+    if (typeof start !== 'number') {
+        throw new TypeError(start + ' must be numeric');
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.splice.js:5]: Same problem here

File: staff/juandavid-omen/arrays/arroz/arroz.prototype.splice.spec.js [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    expect(array.length).toBe(6);
+  });
+
+
+  it("if start is not a number, throw an error", function () {
+    var array = new Arroz(1, 2, 4, 5, 6, 7);
+ 
+     expect(function () {
+      array.splice(true, -3);
+    }).toThrowError(TypeError, 'true must be numeric');
```
Comment (lichab) [inline staff/juandavid-omen/arrays/arroz/arroz.prototype.splice.spec.js:78]: same

**Pull request merged**